tla/ccfraft.tla

```diff
@@ -1013,7 +1013,7 @@ LogMatchingInv ==
 \* of at least one server in every quorum
 QuorumLogInv ==
     \A i \in Servers :
-        \A S \in Quorums[GetServerSetForIndex(i, commitIndex[i])] :
+        \A S \in Quorums[CurrentConfiguration(i)] :
```

Comment:
@heidihoward that does not seem right to me, for reconfiguration transactions, transactions must be in two quorums, one for the old and for the new configuration applicable at that txid. Am I maybe missing something?

Reply:
FYI: Heidi's variant above is violated even with the two related fixes https://github.com/microsoft/CCF/pull/4867 and https://github.com/microsoft/CCF/pull/4868.

Reply:
@achamayou `QuorumLogInv` is an invariant that we believe must be true for all states. It might be weaker than the actual invariant, which it not ideal but perfectly ok. This invariant, which dates back to the OG specification without reconfiguration, is trying to say that if some node believes a prefix of the log to be committed then each quorum knows about that log prefix. What I was trying (albeit unsuccessfully) to express with this patch is that, at a minimum, committed log entries at a given node should be know by the current configuration of the same node, this configuration is both active and after the committed log entries.

Reply:
@lemmy regarding that counterexample, `n4` should have removed `{n1, n2, n5}` from its configurations when it learnt that the next configuration `{n1, n4, n5}` had been committed. This should have been when `n4` updated its commit index to 2 or more. 

In general, when a reconfiguration transaction is committed, all configurations from earlier in the log should be removed from the configurations. The set of configuration is therefore the last committed reconfiguration plus any pending reconfigurations.

Aside, I believe that it is now the case that all terms start with a signature transaction.

Reply:
All sub-actions that change `commitIndex` also change `configurations`, except for sub-action `AppendEntriesAlreadyDone`.  Is this correct?  If yes, we will have to check if the other sub-actions.

Reply:
My understanding is that every time `commitIndex` is updated, including AppendEntriesAlreadyDone, the node should check whether `configurations` needs to be updated accordingly

Reply:
> In general, when a reconfiguration transaction is committed, all configurations from earlier in the log should be removed from the configurations. The set of configuration is therefore the last committed reconfiguration plus any pending reconfigurations.

Yes, this is done at https://github.com/microsoft/CCF/blob/main/src/consensus/aft/raft.h#L2322, inside `commit()`.

> Aside, I believe that it is now the case that all terms start with a signature transaction.

Yes, this happens at https://github.com/microsoft/CCF/blob/main/src/consensus/aft/raft.h#L1912.

> My understanding is that every time commitIndex is updated, including AppendEntriesAlreadyDone, the node should check whether configurations needs to be updated accordingly

Yes, see first item.

Reply:
> > Aside, I believe that it is now the case that all terms start with a signature transaction.
> 
> Yes, this happens at https://github.com/microsoft/CCF/blob/main/src/consensus/aft/raft.h#L1912.
> 

Just to clarifiy, the very first transaction is an exemption to this rule, however, I believe it also cannot be a reconfiguration transaction. This is just an aside, as I do not believe this to be source of the issue

Reply:
@heidihoward the first transaction is unfortunately a little special, yes. It is a configuration transaction in the sense that it writes to the nodes table, a single node (the starter node). It's committable despite being unsigned, among other reasons so that there is always one committed configuration.

Reply:
I'd like get this PR unblocked. 

As @lemmy noted, TLC still finds a violation with the new invariant due to another issue the spec. I've created a seperate ticket to address this issue: https://github.com/microsoft/CCF/issues/4919

Whilst not perfect, I think this change is a step in the right direction.